File: client/csharp/example/connect/test/cli/AcceptMenu.cs
using System;
using System.Collections.Generic;
using Connect;

namespace example
{
  class AcceptMenu {
    private Menu menu;
    private static ConnectSvc connectSvc;

    public AcceptMenu(ConnectSvc svc) {
      connectSvc = svc;

      MenuItem[] items = new MenuItem[6];
      items[0] = new MenuItem("1", "Add devices to the filter", new MenuCallback(AddDevices), false);
      items[1] = new MenuItem("2", "Delete devices from the filter", new MenuCallback(DeleteDevices), false);
      items[2] = new MenuItem("3", "Allow all devices", new MenuCallback(AllowAll), false);
      items[3] = new MenuItem("4", "Disallow all devices", new MenuCallback(DisallowAll), false);
      items[4] = new MenuItem("5", "Refresh the pending device list", new MenuCallback(ShowPendingList), false);
      items[5] = new MenuItem("q", "Return to Main Menu", null, true);      

      menu = new Menu(items);
    }

    public void Show() {
      ShowAcceptFilter();
      ShowPendingList();

      menu.Show("Accept Menu");
    }

    public static void AddDevices() {
      Console.WriteLine("Enter the device IDs to add");

      uint[] deviceIDs = Menu.GetDeviceIDs();

      if(deviceIDs == null) {
        return;
      }

      try {
        AcceptFilter filter = connectSvc.GetAcceptFilter();
        
        for(int i = 0; i < deviceIDs.Length; i++) {
          if(!filter.DeviceIDs.Contains(deviceIDs[i])) {
            filter.DeviceIDs.Add(deviceIDs[i]);
          }
        }

        filter.AllowAll = false;

        connectSvc.SetAcceptFilter(filter);
        ShowAcceptFilter();
      } catch(Exception e) {
        Console.WriteLine("Cannot add the devices to the filter: {0}", e);
      }     
    }

    public static void DeleteDevices() {
      Console.WriteLine("Enter the device IDs to delete");

      uint[] deviceIDs = Menu.GetDeviceIDs();

      if(deviceIDs == null) {
        return;
      }

      try {
        AcceptFilter filter = connectSvc.GetAcceptFilter();
        
        for(int i = 0; i < deviceIDs.Length; i++) {
          filter.DeviceIDs.Remove(deviceIDs[i]);
        }

        filter.AllowAll = false;

        connectSvc.SetAcceptFilter(filter);
        ShowAcceptFilter();
      } catch(Exception e) {
        Console.WriteLine("Cannot delete the devices from the filter: {0}", e);
      }          
    }

    public static void AllowAll() {
      AcceptFilter filter = new AcceptFilter{ AllowAll = true };

      try {
        connectSvc.SetAcceptFilter(filter);
        ShowAcceptFilter();
      } catch(Exception e) {
        Console.WriteLine("Cannot allow all devices: {0}", e);
      }
    }

    public static void DisallowAll() {
      AcceptFilter filter = new AcceptFilter{ AllowAll = false };

      try {
        connectSvc.SetAcceptFilter(filter);
        ShowAcceptFilter();
      } catch(Exception e) {
        Console.WriteLine("Cannot disallow all devices: {0}", e);
      }
    }    

    public static void ShowAcceptFilter() {
      Console.WriteLine("Getting the accept filter...");

      try {
        var filter = connectSvc.GetAcceptFilter();

        Console.WriteLine();
        Console.WriteLine("***** Accept Filter: {0}", filter);
        Console.WriteLine();
      } catch(Exception e) {
        Console.WriteLine("Cannot get the accept filter: {0}", e);
      }
    }

    public static void ShowPendingList() {
      Console.WriteLine("Getting the pending device list...");

      try {
        var devList = connectSvc.GetPendingList();
        
        Console.WriteLine();
        Console.WriteLine("***** Pending Devices: {0}", devList.Count);
        for(int i = 0; i < devList.Count; i++) {
          Console.WriteLine(devList[i]);
        }

        Console.WriteLine();
      } catch(Exception e) {
        Console.WriteLine("Cannot get the pending list: {0}", e);
      }
    }
  }
}

File: client/python/example/connect/test/cli/acceptMenu.py
import grpc

import connect_pb2

from example.cli.menu import MenuItem, Menu
from example.cli.input import getUserInput, UserInput, getDeviceIDs

class AcceptMenu:
  menu = None
  connectSvc = None

  def __init__(self, connectSvc):
    self.connectSvc = connectSvc

    menuItems = [
      MenuItem('1', 'Add devices to the filter', self.addDevicesToFilter, False),
      MenuItem('2', 'Delete devices from the filter', self.deleteDevicesFromFilter, False),
      MenuItem('3', 'Allow all devices', self.allowAllDevices, False),
      MenuItem('4', 'Disallow all devices', self.disallowAllDevices, False),
      MenuItem('5', 'Refresh the pending device list', self.getPendingList, False),
      MenuItem('q', 'Return to Main Menu', None, True),
    ]

    self.menu = Menu('Accept Menu', menuItems)

  def getPendingList(self):
    try:
      devList = self.connectSvc.getPendingList()

      print(f'***** Pending devices: {len(devList)}', flush=True)
      print(devList, flush=True)

      return devList

    except grpc.RpcError as e:
      print(f'Cannot get the pending list: {e}')
      return []

  def showAcceptFilter(self):
    try:
      filter = self.connectSvc.getAcceptFilter()
      print('***** Accept filter: ', filter, flush=True)
    except grpc.RpcError as e:
      print(f'Cannot get the accept filter: {e}')

  def allowAllDevices(self):
    try:
      filter = connect_pb2.AcceptFilter(allowAll=True)
      self.connectSvc.setAcceptFilter(filter)
      self.showAcceptFilter()
    except grpc.RpcError as e:
      print(f'Cannot allow all devices: {e}')      

  def disallowAllDevices(self):
    try:
      filter = connect_pb2.AcceptFilter(allowAll=False)
      self.connectSvc.setAcceptFilter(filter)
      self.showAcceptFilter()
    except grpc.RpcError as e:
      print(f'Cannot disallow all devices: {e}')        

  def addDevicesToFilter(self):
    try:
      deviceIDs = getDeviceIDs()
      if len(deviceIDs) == 0:
        print('No device to add')
        return

      filter = self.connectSvc.getAcceptFilter()
      filter.allowAll = False

      for devID in deviceIDs:
        existing = False
        for existingDevID in filter.deviceIDs:
          if devID == existingDevID:
            existing = True
            break
      
        if not existing:
          filter.deviceIDs.append(devID)

      self.connectSvc.setAcceptFilter(filter)
      self.showAcceptFilter()
          
    except grpc.RpcError as e:
      print(f'Cannot add devices to the filter: {e}')      

  def deleteDevicesFromFilter(self):
    try:
      deviceIDs = getDeviceIDs()
      if len(deviceIDs) == 0:
        print('No device to delete')
        return

      filter = self.connectSvc.getAcceptFilter()
      filter.allowAll = False

      for devID in deviceIDs:
        try:
          filter.deviceIDs.remove(devID)
        except ValueError:
          pass

      self.connectSvc.setAcceptFilter(filter)
      self.showAcceptFilter()
          
    except grpc.RpcError as e:
      print(f'Cannot add devices to the filter: {e}')          

  def show(self):
    self.getPendingList()
    self.showAcceptFilter()
    self.menu.show()